### ISSUE TITLE
False positive "not-context-manager" for psycopg.connect when used as context manager

### ISSUE THREAD

**Issue by tuukkamustonen** [pylint-dev/pylint#5273]
### Bug description

Following:

```py
import psycopg

with psycopg.connect('...') as conn:
    pass
```

...raises E1129 (not-context-manager) violation.

But the code works fine and sources look ok as [psycopg.connect()](https://github.com/psycopg/psycopg/blob/master/psycopg/psycopg/connection.py#L522-L556) returns [Connection](https://github.com/psycopg/psycopg/blob/master/psycopg/psycopg/connection.py#L502) class (which indeed has `__enter__` and `__exit__`).

I'm running `psycopg==3.0.1`.

### Command used

```shell
pylint
```

### Pylint output

```shell
************* Module foo
foo.py:3:0: E1129: Context manager 'NoneType' doesn't implement __enter__ and __exit__. (not-context-manager)

--------------------------------------------------------------------
Your code has been rated at -6.67/10 (previous run: -6.67/10, +0.00)
```

### Expected behavior

There shouldn't be error.

### Pylint version

```shell
pylint 2.11.1
astroid 2.8.4
Python 3.8.12 (default, Nov  6 2021, 14:26:37) 
[GCC 9.3.0]
```

### OS / Environment

Kubuntu 20.04 LTS

**Comment by tuukkamustonen** [comment]
Possibly related: https://github.com/python/mypy/issues/11004

**Comment by rchen152** [comment]
I also ran into this issue. What I discovered is that pylint appears to be checking that the overloads return a contextmanager, which produces an error because overloads return None.  That is:
```
@typing.overload
def some_func():
  ...  # this makes pylint unhappy when it sees `with some_func()` because it looks like `some_func` returns None.

def some_func():
  return MyContextManager()
```

**Comment by rchen152** [comment]
A workaround that worked for me is adding return statements to the overloads:
```
@typing.overload
def some_func():
  return MyContextManager()

[...]
```
Of course, this is rather ugly and only works if you have control over the code containing the overloads.

**Comment by bermed28** [comment]
I also faced issues with this, using `psycopg[binary]` (aka Pscyopg3). I wa using this code per se:

```python
with connect(**connection_params) as conn:
    with conn.cursor() as cursor:
        # some code here
        with cursor.copy(f"COPY {PG_DB_SCHEMA}.{PG_TABLE_NAME} ({columns}) FROM STDIN") as copy:
            # some code here
```

And when running PyLint I got the error:

```shell
************* Module path.to.my.module
         path/to/my/module.py:95:8: E1129: Context manager 'NoneType' doesn't implement __enter__ and __exit__. (not-context-manager)
```

**Comment by Pierre-Sassoulas** [comment]
Sounds great @adamtuft, would you mind adding this test case directly in pylint with a pull request ?

**Comment by adamtuft** [comment]
Sure thing, I'll put one together 👍

**Comment by adamtuft** [comment]
This is the same issue as https://github.com/pylint-dev/pylint/issues/4660 for which you already have a regression test, fixed by https://github.com/pylint-dev/astroid/pull/1173, so I guess there's no need for me to add another test?

**Comment by Pierre-Sassoulas** [comment]
Thanks for checking and sorry for the delay. A regression test in pylint would still be useful (especially if we need to upgrade astroid, it's a way to close this pylint issue only when it's actually fixed in pylint because we upgraded astroid and the test that failed now pass). But of course it's less useful than if we had nothing in astroid, I'll let you judge if you still want to contribute :)

**Comment by jacobtylerwalls** [comment]
@Pierre-Sassoulas I think @adamtuft is pointing out that we actually have this exact case in the test suite, just asserting the wrong result with a sad comment around it. When we upgrade to astroid 3.2 we can just change the result and remove the sad comment 👍.

https://github.com/pylint-dev/pylint/blob/a54d8a68b62d173c872fc1528e16e57453548e2d/tests/functional/r/regression_02/regression_4660.py#L26-L45


Thanks @adamtuft for investigating.

Closing as fixed pending the upgrade to astroid 3.2.

**Comment by freeform-andre** [comment]
I still see this occuring even after picking up the fix
```
$ pylint --version
```
```
pylint 3.3.1
astroid 3.3.5
Python 3.11.0rc1 (main, Aug 12 2022, 10:02:14) [GCC 11.2.0]
```

```python
import psycopg

CONNECTION = psycopg.connect(
    user="admin",
    password="password",
    host="host.com",
    port="1234",
    dbname="important",
    keepalives=1,
)

with CONNECTION as conn:
    with CONNECTION.cursor() as cursor:
        cursor.execute("INSERT INTO tbl(name) VALUES(%s)", ("foobar",))

```
```
$ pylint ./false_positive.py
```
```
************* Module false_positive
false_positive.py:12:0: E1129: Context manager 'NoneType' doesn't implement __enter__ and __exit__. (not-context-manager)

------------------------------------------------------------------
Your code has been rated at 0.00/10 (previous run: 0.83/10, -0.83)
```
